_posts/2023-12-18-homebrew-4.2.0.md

```diff
@@ -0,0 +1,68 @@
+---
+title: 4.2.0
+author: MikeMcQuaid
+redirect_from: /blog/4.2.0/
+---
+
+Today, I'd like to announce Homebrew 4.2.0.
+The most significant changes since 4.1.0 are some major performance upgrades (e.g. using Ruby 3.1, upgrading fewer dependencies), `.env` file configuration and macOS Sonoma support.
+
+Major changes and deprecations since 4.1.0:
+
+- Homebrew now uses and [requires Ruby 3.1](https://github.com/Homebrew/brew/pull/16294).
+  If you do not already have a version provided by your system: we provide
+  [Portable Ruby 3.1.4](https://github.com/Homebrew/brew/pull/16145) that will be installed whenever needed on
+  Linux x86_64 and macOS Apple Silicon and Intel.
+  This marks the end of Homebrew using the macOS system Ruby.
+  Pour one out for our old friend `/usr/bin/ruby` on macOS 🍻.
+- [Formula installation and upgrades are less likely to require dependencies to be upgraded.](https://github.com/Homebrew/brew/pull/15927)
+  This should reduce "`brew upgrade foo` upgraded everything on my system" problems.
+- Homebrew/homebrew-core and Homebrew/homebrew-cask now store formulae/casks in "sharded" subdirectories for improved `git` and GitHub performance.
+- [Homebrew can be configured with `.env` files.](https://github.com/Homebrew/brew/pull/15787)
+- [Homebrew supports macOS Sonoma.](https://github.com/Homebrew/brew/pull/16019)
+- [`OS::Mac` and `MacOS` usage in formulae on Linux is deprecated.](https://github.com/Homebrew/brew/pull/16224)
+  Please guard all uses in formulae with `if OS.mac?` or `if OS.linux?` as appropriate.
```

Comment:
**question:** I'm a few minutes late on this, but are these preferred over `on_macos` and `on_linux` and friends? We're using those inside some internal formulae.

Reply:
Depends. Basic reading of the formulae (i.e. outside `def install/test`) should use `on_{system}` so that you can, e.g., know the dependencies on Linux from a macOS system.

Inside of `def install` only ever runs on the system you are installing on so use `if OS.whatever?` there.

I believe `brew style` checks for the latter case, but does not check for the former case in third-party tap since it doesn't support 100% of use cases. https://github.com/Homebrew/brew/blob/master/Library/Homebrew/rubocops/lines.rb#L403-L442